mev-boost-rs/src/relay_mux.rs

```diff
@@ -124,18 +126,40 @@ impl BlindedBlockProvider for RelayMux {
         &self,
         bid_request: &BidRequest,
     ) -> Result<SignedBuilderBid, BlindedBlockProviderError> {
-        let responses = stream::iter(self.relays.iter().cloned())
-            .map(|relay| async move { relay.fetch_best_bid(bid_request).await })
+        let try_responses = stream::iter(self.relays.iter().cloned())
+            .enumerate()
+            .map(|(relay_index, relay)| async move {
+                let bid = tokio::time::timeout(
+                    std::time::Duration::from_secs(FETCH_BEST_BID_TIME_OUT as u64),
+                    relay.fetch_best_bid(bid_request),
+                )
+                .await;
+                (relay_index, bid)
+            })
             .buffer_unordered(self.relays.len())
             .collect::<Vec<_>>()
             .await;
 
+        let (responses, timed_out): (Vec<_>, Vec<_>) =
+            try_responses.into_iter().partition(|(_, res)| res.is_ok());
+
+        if !timed_out.is_empty() {
+            let timed_out_count = timed_out.len();
+            let relay_indices: Vec<usize> = timed_out.into_iter().map(|(index, _)| index).collect();
+            tracing::warn!(
+                "{} out of {} relays didn't provided bids before {} seconds time out: slow relays: {:?}",
+                timed_out_count,
+                self.relays.len(),
+                FETCH_BEST_BID_TIME_OUT,
+                relay_indices
+            );
+        }
+
         // ideally can fuse the filtering into the prior async fetch but
         // several attempts lead to opaque compiler errors...
         let bids = responses
             .into_iter()
-            .enumerate()
-            .filter_map(|(relay_index, response)| match response {
+            .filter_map(|(relay_index, response)| match response.unwrap() {
                 Ok(mut bid) => {
                     if let Err(err) = validate_bid(&mut bid, &self.context) {
                         tracing::warn!("invalid signed builder bid: {err} for bid: {bid:?}");
```

Comment:
so the aggregate stats in the log ("X of Y relays timed out") but I don't like the use of `partition` because it allocates two additional `Vec`s and we are already seeing this API on the latency-sensitive pathway...

instead, I'd refactor to move the "log a warning upon timeout" inside this `filter_map`

so we can match on `Ok(mut bid)`, `Err(tokio::time::error::Elapsed)`, and finally `Err(err)`

where we can log about the timeout under the `Elapsed` match arm

this avoids having to iterate additional times over the responses (which tbf there are only a small number but this I a domain where we need to be as latency-sensitive as possible)